presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -227,31 +328,34 @@ protected RelationPlan visitJoin(Join node, Void context)
             RelationType right = analysis.getOutputDescriptor(node.getRight());
             List<Expression> leftExpressions = new ArrayList<>();
             List<Expression> rightExpressions = new ArrayList<>();
-            List<ComparisonExpression.Type> comparisonTypes = new ArrayList<>();
+            List<JoinComparisonType> comparisonTypes = new ArrayList<>();
+            List<Optional<Expression>> likeEscapes = new ArrayList<>();
             for (Expression conjunct : ExpressionUtils.extractConjuncts(criteria)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
-                if (!(conjunct instanceof ComparisonExpression)) {
+                if (!isSupportedJoinConjunct(conjunct)) {
                     throw new SemanticException(NOT_SUPPORTED, node, "Unsupported non-equi join form: %s", conjunct);
                 }
 
-                ComparisonExpression comparison = (ComparisonExpression) conjunct;
-                ComparisonExpression.Type comparisonType = comparison.getType();
-                if (comparison.getType() != EQUAL && node.getType() != INNER) {
+                JoinComparisonType comparisonType = JoinComparisonType.fromJoinConjunct(conjunct);
+                if (comparisonType != JoinComparisonType.EQUAL && node.getType() != INNER) {
                     throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins only supported for inner join: %s", conjunct);
                 }
-                Set<QualifiedName> firstDependencies = DependencyExtractor.extractNames(comparison.getLeft(), analysis.getColumnReferences());
```

Comment:
Same comment as in the analyzer. I think we can simplify this and make everything supported if we just divide the conjuncts into two lists: equalities, and everything else. Then everything else will just go into `postInnerJoinComparisons`